File: go/benchmark.go
package main

import (
	"fmt"
	"sync"
	"time"
)

var sharedCounter int
var mutex sync.Mutex

func incrementCounter(wg *sync.WaitGroup) {
	defer wg.Done()

	// Simulate some work
	time.Sleep(time.Duration(time.Now().UnixNano()%100) * time.Millisecond)

	mutex.Lock()
	sharedCounter++
	mutex.Unlock()
}

func runBenchmark(numGoroutines int) {
	sharedCounter = 0
	var wg sync.WaitGroup

	startTime := time.Now()

	for i := 0; i < numGoroutines; i++ {
		wg.Add(1)
		go incrementCounter(&wg)
	}

	wg.Wait()

	endTime := time.Now()

	fmt.Printf("Go - Goroutines: %d, Time: %.4fs, Counter: %d\n", numGoroutines, endTime.Sub(startTime).Seconds(), sharedCounter)
}

func main() {
	for _, numGoroutines := range []int{10, 100, 1000} {
		runBenchmark(numGoroutines)
	}
}

File: python/benchmark.py
import threading
import time
import random

# A shared resource to test thread safety
shared_counter = 0
lock = threading.Lock()

def increment_counter():
    global shared_counter
    # Simulate some work
    time.sleep(random.random() * 0.1)
    with lock:
        shared_counter += 1

def run_benchmark(num_threads):
    global shared_counter
    shared_counter = 0
    threads = []

    start_time = time.time()

    # Create and start threads
    for _ in range(num_threads):
        thread = threading.Thread(target=increment_counter)
        thread.start()
        threads.append(thread)

    # Wait for all threads to complete
    for thread in threads:
        thread.join()

    end_time = time.time()

    print(f"Python - Threads: {num_threads}, Time: {end_time - start_time:.4f}s, Counter: {shared_counter}")

if __name__ == "__main__":
    for num_threads in [10, 100, 1000]:
        run_benchmark(num_threads)